features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,329 @@
-# next webhooks design
+# Background
 
-## Overview
+Web-hooks would be used to notify external services of events in auth gear.
 
-skygear sends webhooks for events that happen in your app, and are designed to help a developer to monitor events, sync database, and many other use cases.
 
-When one of events is triggered, skygear sends a HTTP POST with a payload to the webhook's configured URL.
+# Proposed Designs
 
-## SYNC and ASYNC hooks
+## Events
+Web-hook events are triggered when some mutating operation is performed by
+auth gear.
 
-There are two types of webhook:
+Each operation will trigger two events: BEFORE and AFTER:
+- BEFORE events would be triggered before the operation is performed; the
+  operation can be aborted by web-hook handler.
+- AFTER events would be trigger after the operation is performed.
 
-- `SYNC` - gear's operation will be blocked until a response is received. Gear should respect `SYNC` hook's response, if it indicates failed, operation should abort and return a formatted error response to the client.
-- `ASYNC` - gear's operation won't be blocked and the response from the webhook is ignored.
+Additionally, a `user_sync` event is provided for synchronizing user state to
+self-managed database.
 
-## Request Headers
+Both events has the same event payload.
 
-Following headers will be added to webhook POST request:
 
-- `X-Skygear-Webhook-Req-ID`: an ID of the request.
-- `X-Skygear-Webhook-Async`: if is a `async` hook, then the value is `TRUE` otherwise is `FALSE`. 
-- `X-Skygear-Webhook-Signature`: a signature string of the webhook event, a develop should can verify the request by skygear's SDK or manually.
+## Delivery
 
-## Request Payload
+The events would be delivered to web-hook handlers as an HTTP POST request to
+the handler endpoint.
 
-Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+The handler endpoint can be specified as full URL or an absolute path. Absolute
+paths will be resolved to full URL using an inferred URL scheme and authority:
+- from HTTP request URL if the event is generated from it; otherwise,
+- from the tenant configuration.
 
-All event payloads follows following payload format:
+Each event type can have multiple handlers; the order of deliveries is
+unspecified for AFTER events, and same order as in `skygear.yaml` for BEFORE
+events.
 
-```json=
+BEFORE events will always be delivered before AFTER events. BEFORE events will
+be delivered in well-defined order during a request. AFTER events will be
+delivered in unspecified order.
+
+Web-hook handler should be idempotent, since the same event may be delivered
```

Comment:
Only after event will be delivered multiple times?

Reply:
Currently just after events. However, not sure if we would want to retry before events too in future, so I used 'may'.

Reply:
Let's be consistent first. Either make it clear it is talking about `after webhook` only, OR explain here it is for `after webhook` now but maybe for `before webhook` in future and why... (I suggest no need to explain something we are not sure if we will add in future)